### PR TITLE
Huber Surface Loss: smooth L1 for finer pressure gradient signal

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The baseline uses L1 (MAE) loss for surface pressure. L1 gives **constant gradient magnitude (±1) regardless of error size**. At convergence, most surface nodes have small errors, but L1's constant gradient creates noisy optimization — every node pushes the weights equally hard whether its error is 0.01 or 10.0.

**Huber loss (smooth L1)** transitions from **L2 (quadratic) for small errors** to **L1 (linear) for large errors**:
- Small errors (< δ): gradient ∝ error → smoother, more precise fine-tuning
- Large errors (> δ): gradient = ±1 → robust to outliers (same as current L1)

This gives the optimizer a **finer gradient signal** at convergence while maintaining robustness to extreme values. The transition threshold δ controls the boundary.

**Why this might help:**
1. **Better convergence quality:** At epoch 140+, most surface pressure errors are small. L2-like gradients for small errors enable more precise weight updates, potentially pushing metrics down by 1-3%.
2. **Natural hard-example emphasis:** L2 gives LARGER gradients for larger errors. Nodes near the LE suction peak and separation region (where errors are largest) automatically get stronger gradient signal — a physics-aligned form of hard example mining.
3. **Complementary to existing losses:** The DCT freq loss handles frequency-domain smoothness. `--high_p_clamp` handles extreme outliers. Huber handles the gradient magnitude distribution — a different axis entirely.

**Physical motivation:** Surface pressure distributions have a few high-gradient regions (LE, TE, separation) and large smooth regions (mid-chord). L1 treats them all equally. Huber naturally concentrates optimization effort on the challenging regions.

**Confidence:** Medium. Well-established loss function (Huber 1964, widely used in detection: Faster R-CNN, SSD). The key uncertainty is whether δ is well-tuned — too small → L2 everywhere (sensitive to outliers), too large → L1 everywhere (no benefit).

## Instructions

All changes in `cfd_tandemfoil/train.py`.

### Step 1: Add config flags

```python
huber_surface_loss: bool = False    # use smooth L1 (Huber) instead of L1 for surface pressure
huber_delta: float = 0.5            # transition threshold (in asinh-normalized pressure space)
```

### Step 2: Modify the surface pressure loss computation

Find where the surface pressure MAE loss is computed. It should look something like:
```python
surf_p_loss = F.l1_loss(pred_p[surf_mask], target_p[surf_mask])
# or equivalent using torch.abs(pred - target).mean()
```

Replace with:
```python
if cfg.huber_surface_loss:
    surf_p_loss = F.smooth_l1_loss(
        pred_p[surf_mask], target_p[surf_mask], 
        beta=cfg.huber_delta  # PyTorch uses 'beta' for the Huber threshold
    )
else:
    surf_p_loss = F.l1_loss(pred_p[surf_mask], target_p[surf_mask])
```

**IMPORTANT:** Only change the **surface pressure** loss, NOT:
- Volume losses (keep as-is)
- Velocity losses (keep as-is)
- DCT frequency loss (keep as-is)
- Any other auxiliary losses

The validation metrics should still be computed with L1 (MAE) for fair comparison against baseline.

### Step 3: Choosing delta

The `--huber_delta 0.5` default is calibrated for asinh-normalized pressure space:
- Typical asinh pressure values range from -3 to +3
- Surface MAE ≈ 0.03-0.05 in normalized space for well-predicted nodes
- δ = 0.5 means: errors < 0.5 get L2 treatment, errors > 0.5 get L1 treatment
- This puts most nodes in the L2 regime (fine-tuning) while keeping outlier robustness

If the default doesn't work, try δ = 0.1 (more L2, more aggressive fine-tuning) or δ = 1.0 (more L1, closer to baseline).

### Step 4: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent askeladd --wandb_name "askeladd/huber-surface-s42" \
  --wandb_group "round18/huber-surface-loss" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --huber_surface_loss --huber_delta 0.5

# Seed 73 — identical but --seed 73 --wandb_name "askeladd/huber-surface-s73"
```

### Step 5: Report results

Table: p_in, p_oodc, p_tan, p_re for both seeds, 2-seed avg, baseline comparison, W&B run IDs.

**Also report:** 
- Training loss curve shape (should be lower than baseline L1 since Huber < L1 for small errors)
- Whether the validation MAE (still computed as L1) improves even though training uses Huber

## Baseline

Current best (PR #2213, Wake Deficit Feature, 2-seed average):

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| **p_tan** | **28.341** | **< 28.34** |
| p_re   | **6.300**  | < 6.30  |

W&B runs: `hgml7i2r` (seed 42), `qic03vrg` (seed 73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent askeladd --wandb_name "askeladd/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Results — Huber Surface Loss

### Summary

Replaced L1 (MAE) with smooth L1 (Huber) loss for surface pressure training only, with δ=0.5. Validation metrics still computed with L1 for fair comparison. Only the `_surf_pres_err` tensor was changed — volume losses, velocity losses, DCT freq loss all unchanged.

### Metrics

| Metric | Baseline (PR #2213) | S42 | S73 | 2-seed avg | Delta |
|--------|---------------------|-----|-----|------------|-------|
| p_in   | **11.979** | 15.0 | 20.8 | 17.90 | **+49%** |
| p_tan  | **28.341** | 28.9 | 30.9 | 29.90 | **+5.5%** |
| p_oodc | **7.643** | 10.8 | 12.2 | 11.50 | **+50%** |
| p_re   | **6.300** | 8.2 | 9.9 | 9.05 | **+44%** |

**W&B runs:** `3onbhuby` (seed 42), `kn5v2ib6` (seed 73)

### Analysis

- **Clear negative result.** All metrics significantly worse, especially p_in (+49%) and p_oodc (+50%).
- **Seed 73 much worse than S42** — high variance between seeds suggests the Huber loss creates a less stable training landscape.
- **Root cause hypothesis:** δ=0.5 is likely too large for the asinh-normalized pressure space. With `asinh_scale=0.75`, typical normalized pressure errors at convergence are quite small (0.01-0.05). A δ=0.5 means essentially ALL surface nodes are in the L2 regime, which:
  1. **Reduces gradient magnitude** for small errors: L2 gives gradient ∝ error, while L1 gives gradient = 1. At convergence, most errors are small → Huber gradients are much weaker → slower/worse optimization.
  2. **Interacts badly with the surface weight scaling**: The adaptive surf_weight (5-50x multiplier) was tuned for L1-scale gradients. With Huber's smaller gradients, the effective surface loss contribution drops substantially.
  3. **Hard-node mining threshold shifts**: The hard-node mining at epoch 30+ uses median of errors to set threshold. Huber errors are smaller than L1 errors, shifting the threshold and the mining behavior.

### Suggested follow-ups

1. **Much smaller δ (0.01-0.05)**: This would keep most nodes in L1 regime (matching baseline behavior) while smoothing only the smallest errors near zero. δ=0.01 is probably the right scale for asinh-normalized pressure.
2. **Scale compensation**: Multiply the Huber loss by 2/δ to match L1 gradient scale at large errors, preserving the surf_weight balance.
3. **L1 + L2 combination**: Instead of Huber, try `(1-α)*L1 + α*L2` with small α (0.05-0.1). This preserves L1's constant gradients while adding a small L2 component for fine-tuning.
4. **Only apply Huber late in training**: Use L1 for first 100 epochs, then switch to Huber for final fine-tuning. This avoids the weak-gradient problem during early training.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2236 because ADVISOR: Closing — severe regression across all metrics (p_in +49%, p_oodc +50%, p_re +44%, p_tan +5.5%). Your analysis is correct: δ=0.5 puts essentially all nodes in L2 regime where gradient ∝ error. With small errors at convergence, this dramatically weakens the gradient signal. The interaction with surf_weight scaling (tuned for L1-scale gradients) and hard-node mining thresholds makes this particularly destructive. While a smaller δ might partially address this, the fundamental issue is that the entire training pipeline (surf_weight, high_p_clamp, PCGrad) is carefully co-tuned for L1 gradients — swapping the loss function requires re-tuning everything downstream. Thank you for the thorough analysis..
